cub/cub/detail/rfa.cuh

```diff
@@ -0,0 +1,1258 @@
+#  include <array>
+using std::array;
+using std::max;
+using std::min;
+#else
+#  include <cuda/std/__algorithm_>
+#  include <cuda/std/array>
+using cuda::std::array;
+using cuda::std::max;
+using cuda::std::min;
+
+// disable zero checks
+#  define DISABLE_ZERO
+
+// disable nan / infinity checks
+#  define DISABLE_NANINF
+
+// jump table for indexing into data
+#  define MAX_JUMP 5
+static_assert(MAX_JUMP <= 5, "MAX_JUMP greater than max");
+
+#endif
+
+CUB_NAMESPACE_BEGIN
+
+namespace detail
+{
+namespace rfa_detail
+{
+
+template <typename Float, int LEN>
+static __device__ Float* get_shared_bin_array()
+{
+  static __shared__ Float bin_computed_array[LEN];
+  return bin_computed_array;
+}
+
+template <class T>
+struct get_vector_type
+{};
+template <>
+struct get_vector_type<double>
+{
+  using type = double2;
+};
+template <>
+struct get_vector_type<float>
+{
+  using type = float4;
+};
+
+template <class T>
+using vector_t = typename get_vector_type<T>::type;
+
+template <class T>
+std::size_t vector_size()
+{
+  return sizeof(vector_t<T>) / sizeof(T);
+}
+
+template <class T>
+inline static auto abs_max(const T&);
+
+template <>
+__host__ __device__ auto abs_max(const float4& x)
+{
+  return fmax(fmaxf(fabs(x.x), fabs(x.y)), fmax(fabs(x.z), fabs(x.w)));
+}
+
+template <>
+__host__ __device__ auto abs_max(const double4& x)
+{
+  return fmax(fmax(fabs(x.x), fabs(x.y)), fmax(fabs(x.z), fabs(x.w)));
+}
+
+template <>
+__host__ __device__ auto abs_max(const double2& x)
+{
+  return fmax(fabs(x.x), fabs(x.y));
+}
+
+template <>
+__host__ __device__ auto abs_max(const float2& x)
+{
+  return fmax(fabs(x.x), fabs(x.y));
+}
+
+template <class ftype>
+struct RFA_bins
+{
+  static constexpr auto BIN_WIDTH = std::is_same_v<ftype, double> ? 40 : 13;
+  static constexpr int MIN_EXP    = std::numeric_limits<ftype>::min_exponent;
+  static constexpr int MAX_EXP    = std::numeric_limits<ftype>::max_exponent;
+  static constexpr auto MANT_DIG  = std::numeric_limits<ftype>::digits;
+  /// Binned floating-point maximum index
+  static constexpr auto MAXINDEX = ((MAX_EXP - MIN_EXP + MANT_DIG - 1) / BIN_WIDTH) - 1;
+  // The maximum floating-point fold supported by the library
+  static constexpr auto MAXFOLD = MAXINDEX + 1;
+
+  __device__ __host__ static ftype initialize_bins(int index)
+  {
+    if (index == 0)
+    {
+      return ldexpf(0.75f, MAX_EXP);
+    }
+    else
+    {
+      return ldexpf(0.75f, MAX_EXP + MANT_DIG - BIN_WIDTH - index * BIN_WIDTH);
+    }
+  }
+};
+
+/// Class to hold a reproducible summation of the numbers passed to it
+///
+///@param ftype Floating-point data type; either `float` or `double
+///@param FOLD  The fold; use 3 as a default unless you understand it.
+template <class ftype_, int FOLD_ = 3, typename std::enable_if_t<std::is_floating_point<ftype_>::value>* = nullptr>
+class alignas(2 * sizeof(ftype_)) ReproducibleFloatingAccumulator
+{
+public:
+  using ftype = ftype_;
+  enum
+  {
+    FOLD = FOLD_
```

Comment:
I do not see a reason for the enum. Could you elaborate why a template argument wont suffice?

Reply:
There was an instance where `MAXINDEX` was defined as a `constexpr` variable and that caused an error and I honestly don't remember why I defined `FOLD` as an enum instead of just a template parameter. Sorry, will use it as one and revert back if such an instance occurs again. 

Thanks!